### PR TITLE
Windows compatibility

### PR DESCRIPTION
**Pull request by mehrdadn**
Contains changes required for Windows version to build correctly. (No tests have been run yet.)

### CODE REVIEW

File: src/plasma/plasma_client.c [hunk 1]
```diff
@@ -333,14 +342,18 @@ int64_t plasma_evict(plasma_connection *conn, int64_t num_bytes) {
 
 int plasma_subscribe(plasma_connection *conn) {
   int fd[2];
+  /* TODO: Just create 1 socket,
+   * bind it to port 0 to find a free port,
+   * and send the port number instead, and let the client connect.
+   */
```
Comment (robertnishihara) [inline src/plasma/plasma_client.c]: The communication over the port could have different performance though, right? Couldn't it be slower?

**Review by robertnishihara — COMMENTED** [review]
Nice, this looks pretty good to me!

I have a few questions (left as comments).

I have some minor style comments, but I can go through and do those myself.

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[submodule "src/common/thirdparty/redis"]
+	path = src/common/thirdparty/redis-windows
+	url = https://github.com/MSOpenTech/redis.git
+	ignore = all
+[submodule "src/common/thirdparty/python"]
+	path = src/common/thirdparty/python
+	url = https://github.com/austinsc/python.git
+	ignore = all
```
Comment (mehrdadn) [inline .gitmodules]: I assume you mean `git clone` and not `wget`...
It's of course possible... but it just makes life hard. Why?

File: src/common/common.h [hunk 3]
```diff
@@ -63,6 +65,13 @@
     backtrace_symbols_fd(buffer, calls, 1);                               \
     exit(-1);                                                             \
   } while (0);
+#else
+#define LOG_FATAL(M, ...)                                          \
+  do {                                                             \
+    fprintf(stderr, "[FATAL] (%s:%d) " M "\n", __FILE__, __LINE__, \
+            ##__VA_ARGS__);                                        \
+    exit(-1);                                                      \
+  } while (0);
```
Comment (mehrdadn) [inline src/common/common.h]: Yeah we don't, I just copy-pasted.

File: src/common/lib/python/common_extension.c [hunk 4]
```diff
@@ -99,7 +99,7 @@ static long PyObjectID_hash(PyObjectID *self) {
 }
 
 static PyObject *PyObjectID_repr(PyObjectID *self) {
-  int hex_length = 2 * UNIQUE_ID_SIZE + 1;
+  enum { hex_length = 2 * UNIQUE_ID_SIZE + 1 };
```
Comment (mehrdadn) [inline src/common/lib/python/common_extension.c]: Non-constant array sizes aren't valid in pre-C99(?) and gave me errors. And they don't compile to the same code post-C99.

File: src/common/shims/windows/getopt.h [hunk 5]
```diff
@@ -0,0 +1 @@
+#pragma once
```
Comment (mehrdadn) [inline src/common/shims/windows/getopt.h]: Feel free to change it.

**Comment by raopku** [comment]
can ray be installed on windows ?

**Comment by mehrdadn** [comment]
Unfortunately not currently. You may want to track issue #631 regarding this.

**Pull request merged**